test/integration/ddl_sharding_info_reload_test.lua

```diff
@@ -0,0 +1,118 @@
+    g.cluster:start()
+end)
+
+pgroup_new_space.after_all(function(g)
+    helpers.stop_cluster(g.cluster)
+end)
+
+pgroup_new_space.before_each(function(g)
+    helpers.drop_space_on_cluster(g.cluster, 'customers_new')
+
+    helpers.call_on_storages(g.cluster, function(server)
+        server.net_box:call('reset_to_default_schema')
+    end)
+
+    -- Assert space doesn't exist.
+    local obj, err = g.cluster.main_server.net_box:call(
+       'crud.insert', {'customers_new', {1, box.NULL, 'Emma', 22}}
+    )
+
+    t.assert_equals(obj, nil)
+    t.assert_is_not(err, nil)
+    t.assert_str_contains(err.err, "Space \\-\"customers_new\\-\" doesn't exist", true)
+end)
+
+
+-- Test metainfo is updated on router if new space added to ddl.
+
+local test_tuple = {1, box.NULL, 'Emma', 22}
+local test_object = { id = 1, bucket_id = box.NULL, name = 'Emma', age = 22 }
+
+-- Sharded by "name" and computed with custom sharding function.
+local test_customers_new_result = {
+    s1 = {1, 2862, 'Emma', 22},
+    s2 = nil,
+}
+
+local new_space_cases = {
+    insert = {
+        func = 'crud.insert',
+        input = {'customers_new', test_tuple},
+        result = test_customers_new_result,
+    },
+    insert_object = {
+        func = 'crud.insert_object',
+        input = {'customers_new', test_object},
+        result = test_customers_new_result,
+    },
+    replace = {
+        func = 'crud.replace',
+        input = {'customers_new', test_tuple},
+        result = test_customers_new_result,
+    },
+    replace_object = {
+        func = 'crud.replace_object',
+        input = {'customers_new', test_object},
+        result = test_customers_new_result,
+    },
+    upsert = {
+        func = 'crud.upsert',
+        input = {'customers_new', test_tuple, {}},
+        result = test_customers_new_result,
+    },
+    upsert_object = {
+        func = 'crud.upsert_object',
+        input = {'customers_new', test_object, {}},
+        result = test_customers_new_result,
+    },
+}
+
+for name, case in pairs(new_space_cases) do
+    local test_name = ('test_%s'):format(name)
+
+    pgroup_new_space[test_name] = function(g)
+        -- Create space 'customers_new', sharded by 'name'.
+        helpers.call_on_storages(g.cluster, function(server)
+            server.net_box:call('create_new_space')
+        end)
+
+        -- Assert it is now possible to call opertions for a new space.
+        local obj, err = g.cluster.main_server.net_box:call(case.func, case.input)
+        t.assert_is_not(obj, nil)
+        t.assert_equals(err, nil)
+
+        -- Assert it is sharded based on updated ddl info.
+        local conn_s1 = g.cluster:server('s1-master').net_box
+        local result = conn_s1.space['customers_new']:get(1)
+        t.assert_equals(result, case.result.s1)
+
+        local conn_s2 = g.cluster:server('s2-master').net_box
+        local result = conn_s2.space['customers_new']:get(1)
+        t.assert_equals(result, case.result.s2)
+    end
+end
```

Comment:
This test imposes relaxed guarantees comparing to what is stated in the commit message. The commit message says:

> Add tests to validate that if space was added at runtime, its sharding
info is fetched to router.

But look at the code (from `master`):

https://github.com/tarantool/crud/blob/d1b1c1aa15388fde282f4ee142333ca1c3907ad4/crud/common/sharding/sharding_metadata.lua#L135-L138

If we have *some* fetched metainfo, we'll use it. If the space is not present here, we'll use the default `strcrc32()` on the primary key.

What is going on here:

1. Before a first request we have no schema, so it'll be fetched. Here the test already incorrect (according to its definition in the commit message): we should perform some request to trigger schema fetch, then add a new space, then issue a request against it.
2. Next, we going to the next test case. The space name in it is the same as in the first one, so we'll just use old cached info. The test cases should use different space names.

Maybe I misunderstood a goal of the commit 'ddl: test new space sharding info is fetched' (I'm in progress of reading the patchset and didn't look at the following patches) and here we want to add a stub to show what is working before next commits? If so, it is not clear from the description of the commit.

Reply:
Thank you for noticing! In final commit reordering session I messed some things up, so let's clarify them and fix tests.

> Before a first request we have no schema, so it'll be fetched. Here the test already incorrect (according to its definition in the commit message): we should perform some request to trigger schema fetch, then add a new space, then issue a request against it.

I perform a request for non-existing in `before_each` hook, so schema must be fetched before each test case. What I forgot is that to use outdated ddl schema we must have some non-null ddl schema before space is created. I've added a non-empty ddl schema fetch and tests are failing now. I reworked them to be similar to "migration" tests: check if we get hash mismatch error in "ddl: fail on sharding info mismatch" commit and check that info is reloaded in "ddl: reload and retry on sharding info mismatch" commit.

> Next, we going to the next test case. The space name in it is the same as in the first one, so we'll just use old cached info. The test cases should use different space names.

Again, since we get an explicit "space doesn't exist" error in before_test hook, space schema must be updated. Or am I missing something?

Reply:
> Again, since we get an explicit "space doesn't exist" error in before_test hook, space schema must be updated. Or am I missing something?

Not in current `master` if 'schema' assumes data from `_ddl_sharding_*` spaces. But since the test is now goes after your changes, it is irrelevant.